Src/VimCore/UndoRedoOperations.fs

```diff
@@ -363,7 +363,16 @@ and UndoRedoOperations
                 | UndoRedoData.Linked (count, false) :: tail -> 
                     Contract.Assert(count > 0)
                     hasData <- true
-                    _undoStack <- UndoRedoData.Linked (count, true) :: tail
+
+                    // Append an empty transaction to the previous undo transaction
+                    // to break the undo sequence.
+                    match _textUndoHistory with
+                    | Some textUndoHistory ->
+                        use undoTransaction = textUndoHistory.CreateTransaction "Break undo sequence"
+                        undoTransaction.Complete()

```

Comment:
Putting the fix here means that every single linked undo transaction now has an extra transaction put on top of it. That feels like a pretty big hammer. Won't that impact all of the other places, in addition to the insert case, where we use linxed undo transactions? I'm thinking of all the cases in command util where this happens.

Reply:
I have been testing this for a few days with no ill effects, but I agree now that the fix is fairly invasive for the how seemingly minor the bug is (that `<C-g>u` in insert mode is useless).

I also came across a very helpful way to visualize the undo stack using the drop down on the undo button in the toolbar:

![image](https://user-images.githubusercontent.com/2680524/41888222-5dcf2de8-78d2-11e8-94ec-0dd119bd1f10.png)

Anyway, to address your comment, I think we only need to add the empty undo transact after inserts, which will cut down on the number of undo transactions considerably. I'll make that change.

Reply:
Hmm. After debugging into this a bit I understand what is going on here better. The native VS undo manager, which is what backs undo in the VS case, will create a single transaction for text edits which can be merged so long as the text edit is forward. I'm a bit worried the current implementation of undo doesn't fully account for this. Going to chat with someone tomorrow about this.

Reply:
Yes, the problem does not manifest in either the unit tests or VimApp because of the simpler (non-merging) undo implementation. This is why I didn't catch it when I implemented `<C-g>u` in insert mode to break the undo sequence.

My original goal, after understanding that some undo transactions can be merged, was to somehow create an unmergeable insert transaction. That seems doable, except it turned out that Visual Studio itself is creating the insert transactions. Even if we could get it work, it gets into implementing our own undo transactions, undo primitives, merge strategies, etc. which itself is a lot of work when what we want is simple: please don't merge these adjacent insert transactions.

Reply:
Here is the a sample of the undo stack with the new "append empty undo transaction only when the linked transaction ends with an insert". I did one insert (`aabc<Esc>`), then another insert (`adef<Esc>`), then a normal command that modified the buffer (`:sort`):

![image](https://user-images.githubusercontent.com/2680524/41891922-52c80e62-78e4-11e8-90fe-9228b3f4d390.png)

As you can see, the normal command (which creates a linked undo transaction), does not cause the empty undo transaction to be appended.